File: problems/valid_palindrome/valid_palindrome.py
def isPalindrome(s: str) -> bool:
    left, right = 0, len(s) - 1
    s = s.lower()

    while left < right:
        s_l, s_r = s[left], s[right]

        if not s_l.isalnum():
            left += 1
            continue

        if not s_r.isalnum():
            right -= 1
            continue

        if s_l != s_r:
            return False

        left += 1
        right -= 1
    return True

File: problems/valid_palindrome/valid_palindrome.cc
#include "valid_palindrome.h"

#include <algorithm>

using std::string;

bool isPalindrome(string s) {
  int left = 0;
  int right = static_cast<int>(s.size()) - 1;

  // Transform the string (in-place) to lowercase
  std::transform(s.begin(), s.end(), s.begin(),
                 [](unsigned char c) { return std::tolower(c); });

  while (left < right) {
    // Move left pointer to the next alphanumeric character
    while (left < right && !std::isalnum(s[left])) {
      ++left;
    }

    // Move right pointer to the previous alphanumeric character
    while (left < right && !std::isalnum(s[right])) {
      --right;
    }

    // Compare characters at left and right pointers
    if (s[left] != s[right]) {
      return false; // Not a palindrome
    }

    ++left;
    --right;
  }
  return true; // Is a palindrome
}